src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.cpp

```diff
@@ -93,19 +105,42 @@ namespace
         {
         }
 
-        double convert_density(double pdf, const BDPTVertex& vertex) const
+        // Returns the density converted from probability per steridian to probability per unit area (Veach 8.2.2.2).
+        float convert_density(double pdf, const BDPTVertex& next) const
         {
-            const Vector3d w = m_position - vertex.m_position;
+            const Vector3d w = next.m_position - m_position;
             const double dist2 = square_norm(w);
             if (dist2 == 0.0)
                 return 0.0;
             const double rcp_dist2 = 1.0 / dist2;
-            pdf *= max(dot(vertex.m_geometric_normal, w * sqrt(rcp_dist2)), 0.0);
-            return pdf * rcp_dist2;
+            pdf *= std::abs(dot(next.m_geometric_normal, w * sqrt(rcp_dist2)));
+            return static_cast<float>(pdf * rcp_dist2);
```

Comment:
`pdf` has a `sqrt(rcp_dist2)` term in numerator, maybe we can simplify and simply divide by `sqrt(rcp_dist2)` here (and remove the term from the numerator)?